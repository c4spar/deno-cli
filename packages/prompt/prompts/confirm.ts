import { blue, bold, dim, yellow } from 'https://deno.land/std@v0.41.0/fmt/colors.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputPromptOptions, GenericInputPromptSettings } from '../lib/generic-input.ts';

export interface ConfirmOptions extends GenericInputPromptOptions<boolean> {
    active?: string;
    inactive?: string;
}

export interface ConfirmSettings extends GenericInputPromptSettings<boolean> {
    active: string;
    inactive: string;
}

export class Confirm extends GenericInput<boolean, ConfirmSettings> {

    public static async prompt( options: string | ConfirmOptions ): Promise<boolean | undefined> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            active: 'y',
            inactive: 'n',
            pointer: blue( Figures.POINTER_SMALL ),
            ...options
        } ).prompt();
    }

    protected getMessage(): string {

        let message = ` ${ yellow( '?' ) } ${ bold( this.settings.message ) } `;

        if ( this.settings.default === true ) {
            message += dim( `(${ this.settings.active.toUpperCase() }/${ this.settings.inactive.toLowerCase() })` );
        } else if ( this.settings.default === false ) {
            message += dim( `(${ this.settings.active.toLowerCase() }/${ this.settings.inactive.toUpperCase() })` );
        } else {
            message += dim( `(${ this.settings.active.toLowerCase() }/${ this.settings.inactive.toLowerCase() })` );
        }

        return message;
    }

    protected transform( value: string ): boolean | undefined {

        switch ( value.toLowerCase() ) {

            case 'y':
            case 'yes':
                return true;

            case 'n':
            case 'no':
                return false;
        }

        return;
    }

    protected validate( value: boolean | undefined ): boolean {
        return typeof value === 'boolean';
    }

    protected format( value: boolean ): string {
        return value ? 'Yes' : 'No';
    }
}
