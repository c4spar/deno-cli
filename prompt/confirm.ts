import { blue, bold, dim, yellow } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic_input.ts";

export type ConfirmKeys = GenericInputKeys;

/** Confirm prompt options. */
export interface ConfirmOptions extends GenericInputPromptOptions<boolean> {
  active?: string;
  inactive?: string;
  keys?: ConfirmKeys;
}

/** Confirm prompt settings. */
interface ConfirmSettings extends GenericInputPromptSettings<boolean> {
  active: string;
  inactive: string;
  keys?: ConfirmKeys;
}

/** Confirm prompt representation. */
export class Confirm extends GenericInput<boolean, ConfirmSettings> {
  /** Execute the prompt and show cursor on end. */
  public static prompt(
    options: string | ConfirmOptions,
  ): Promise<boolean> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      active: "Yes",
      inactive: "No",
      pointer: blue(Figures.POINTER_SMALL),
      ...options,
    }).prompt();
  }

  protected defaults(): string {
    let defaultMessage = "";

    if (this.settings.default === true) {
      defaultMessage += this.settings.active[0].toUpperCase() + "/" +
        this.settings.inactive[0].toLowerCase();
    } else if (this.settings.default === false) {
      defaultMessage += this.settings.active[0].toLowerCase() + "/" +
        this.settings.inactive[0].toUpperCase();
    } else {
      defaultMessage += this.settings.active[0].toLowerCase() + "/" +
        this.settings.inactive[0].toLowerCase();
    }

    return defaultMessage ? dim(` (${defaultMessage})`) : "";
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    return typeof value === "string" &&
      [
          this.settings.active[0].toLowerCase(),
          this.settings.active.toLowerCase(),
          this.settings.inactive[0].toLowerCase(),
          this.settings.inactive.toLowerCase(),
        ].indexOf(value.toLowerCase()) !== -1;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): boolean | undefined {
    switch (value.toLowerCase()) {
      case this.settings.active[0].toLowerCase():
      case this.settings.active.toLowerCase():
        return true;
      case this.settings.inactive[0].toLowerCase():
      case this.settings.inactive.toLowerCase():
        return false;
    }
    return;
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: boolean): string {
    return value ? this.settings.active : this.settings.inactive;
  }
}
