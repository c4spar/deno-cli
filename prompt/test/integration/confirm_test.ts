import { Confirm } from "../../confirm.ts";
import { assertSnapshotCall } from "../../../testing/assert_snapshot_call.ts";

await assertSnapshotCall({
  name: "confirm prompt",
  meta: import.meta,
  osSuffix: ["windows"],
  // args: ["--allow-env=ASSERT_SNAPSHOT_CALL_TEST_NAME"],
  steps: {
    "should confirm": { stdin: ["y", "\n"] },
    "should not confirm": { stdin: ["n", "\n"] },
    "should not confirm by default": { stdin: ["", "\n"] },
  },
  async fn() {
    await Confirm.prompt({
      message: "Please confirm",
      hint: "some hint",
      default: false,
    });
  },
});
