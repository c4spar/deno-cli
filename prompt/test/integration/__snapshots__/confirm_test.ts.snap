export const snapshot = {};

snapshot[`confirm prompt > should confirm 1`] = `
stdout:
--------------------------
? Please confirm (y/N) ‚Ä∫ 
‚ùØ some hint[1A[26G[G[0J
? Please confirm (y/N) ‚Ä∫ yes
‚ùØ some hint[1A[27G[G[0J
? Please confirm (y/N) ‚Ä∫ Yes
[?25h[?25h
--------------------------
stderr:
--------------------------
--------------------------
`;

snapshot[`confirm prompt > should not confirm 1`] = `
stdout:
--------------------------
? Please confirm (y/N) ‚Ä∫ 
‚ùØ some hint[1A[26G[G[0J
? Please confirm (y/N) ‚Ä∫ no
‚ùØ some hint[1A[27G[G[0J
? Please confirm (y/N) ‚Ä∫ No
[?25h[?25h
--------------------------
stderr:
--------------------------
--------------------------
`;

snapshot[`confirm prompt > should not confirm by default 1`] = `
stdout:
--------------------------
? Please confirm (y/N) ‚Ä∫ 
‚ùØ some hint[1A[26G[G[0J
? Please confirm (y/N) ‚Ä∫ No
[?25h[?25h
--------------------------
stderr:
--------------------------
--------------------------
`;
